OpenRA.Mods.Common/Traits/Power/Player/PowerManager.cs

```diff
@@ -45,59 +45,38 @@ public PowerManager(Actor self, PowerManagerInfo info)
 			this.self = self;
 			this.info = info;
 
-			self.World.ActorAdded += UpdateActor;
-			self.World.ActorRemoved += RemoveActor;
-
 			devMode = self.Trait<DeveloperMode>();
 			wasHackEnabled = devMode.UnlimitedPower;
 		}
 
 		public void UpdateActor(Actor a)
 		{
-			UpdateActors(new[] { a });
-		}
-
-		public void UpdateActors(IEnumerable<Actor> actors)
-		{
-			foreach (var a in actors)
-			{
-				if (a.Owner != self.Owner)
-					return;
-
-				var power = a.TraitOrDefault<Power>();
-				if (power == null)
-					return;
-
-				powerDrain[a] = power.GetCurrentPower();
-			}
-			UpdateTotals();
+			int old;
+			powerDrain.TryGetValue(a, out old); // old is 0 if a is not in powerDrain
+			var amount = a.TraitsImplementing<Power>().Where(t => !t.IsTraitDisabled).Aggregate(0, (v, p) => v + p.GetEnabledPower());
+			powerDrain[a] = amount;
+			if (amount == old || devMode.UnlimitedPower)
+				return;
+			if (old > 0)
+				totalProvided -= old;
+			else if (old < 0)
+				totalDrained += old;
+			if (amount > 0)
+				totalProvided += amount;
+			else if (amount < 0)
+				totalDrained -= amount;
 		}
 
-		void RemoveActor(Actor a)
+		public void RemoveActor(Actor a)
 		{
-			if (a.Owner != self.Owner || !a.HasTrait<Power>())
+			int amount;
+			if (!powerDrain.TryGetValue(a, out amount))
 				return;
-
+			if (amount > 0)
+				totalProvided -= amount;
+			else if (amount < 0)
+				totalDrained += amount;
 			powerDrain.Remove(a);
-			UpdateTotals();
-		}
-
-		public void UpdateTotals()
-		{
-			totalProvided = 0;
-			totalDrained = 0;
-
-			foreach (var kv in powerDrain)
-			{
-				var p = kv.Value;
-				if (p > 0)
-					totalProvided += p;
-				else
-					totalDrained -= p;
-			}
-
-			if (devMode.UnlimitedPower)
-				totalProvided = 1000000;
```

Comment:
You made this essentially not `UnlimitedPower` but `DisablePowerDrain` and accidentally broke things. https://github.com/OpenRA/OpenRA/issues/7317